PCHR-3417: Delete unnecesary activity types

reneolivo opened the pull request:
## Overview
This PR deletes activity types that are not needed by the extension.

## API Call
Some of the activity types to be deleted:
```php
civicrm_api3('OptionValue', 'get', array(
  'sequential' => 1,
  'option_group_id' => "activity_type",
  'name' => array('IN' => array("Downloaded Invoice", "Emailed Invoice", "Inbound SMS", "Outbound SMS", "SMS delivery", "Tell a Friend")),
));
```

## Before

```json
{
    "is_error": 0,
    "version": 3,
    "count": 5,
    "values": [ ... ]
}
```

## After

```json
{
    "is_error": 0,
    "version": 3,
    "count": 0,
    "values": [

    ]
}
```


## Technical Details

The upgrader code is defined as this:

```php
public function upgrade_1035() {
  // removes all activities from these components:
  $this->_uninstallAllComponentActivities([
    'CiviCampaign',
    'CiviContribute',
    'CiviEvent',
    'CiviMember',
    'CiviPledge',
  ]);

  // removes activities using their label as reference:
  $this->_uninstallActivitiesByLabel([
    'Downloaded Invoice',
    'Emailed Invoice',
    'Inbound SMS',
    'Outbound SMS',
    'SMS delivery',
    'Tell a Friend',
  ]);

  // removes activities from the CiviCase component, using their name as reference:
  $this->_uninstallActivityTypes('CiviCase', [
    'Background Check',
    'Collate and print goals',
    'Collection of Appraisal forms',
    'Collection of appraisal paperwork',
    'Conduct appraisal',
    'Follow up on progress',
    'ID badge',
    'Interview Prospect',
    'Issue confirmation/warning letter',
    'Issue extension letter',
    'Prepare and email schedule',
    'Prepare formats',
    'Print formats',
    'Revoke access to databases',
  ]);

  // removes activities from the CiviDocument component, using their name as reference:
  $this->_uninstallActivityTypes('CiviDocument', [
    'Joining Document 2',
    'Joining Document 3',
    'Exiting document 2',
    'Exiting document 3',
  ]);

  return TRUE;
}
```

The upgrader uses the existing `_uninstallActivityTypes` function, though minor syntax refactoring was done to it: https://github.com/compucorp/civihr-tasks-assignments/blob/7d47c09f4cfe7cef8ad994cdb3bebd20eb123b75/uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php#L942-L954

To remove activities from a list of components `_uninstallAllComponentActivities` was defined:

```php
private function _uninstallAllComponentActivities(array $componentNames) {
  $result = civicrm_api3('OptionValue', 'get', [
    'sequential' => 1,
    'option_group_id' => 'activity_type',
    'component_id' => [ 'IN' => $componentNames ],
    'return' => 'id'
  ]);

  foreach ($result['values'] as $activity) {
    civicrm_api3('OptionValue', 'delete', ['id' => $activity['id']]);
  }
}
```

And to remove activities using their label as reference `_uninstallActivitiesByLabel` was defined:

```php
private function _uninstallActivitiesByLabel(array $activityLabels) {
  $result = civicrm_api3('OptionValue', 'get', [
    'sequential' => 1,
    'option_group_id' => 'activity_type',
    'label' => [ 'IN' => $activityLabels ],
    'return' => 'id'
  ]);

  foreach ($result['values'] as $activity) {
    civicrm_api3('OptionValue', 'delete', ['id' => $activity['id']]);
  }
}
```

Removing these activities by label instead of name was done since their label is more precise (Ex: `Inbound SMS` and `Outbound SMS` labels vs their `SMS` name). Also, these activities don't belong to a particular extension.


## Comments

* A `return TRUE` was missing from the upgrader introduced in https://github.com/compucorp/civihr-tasks-assignments/pull/334 This PR adds that missing statement.
* Lint issues were fixed for the whole upgrader file.

mickadoo commented on uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php:
```diff
@@ -756,15 +794,66 @@ public function upgrade_1033() {
   }
 
+  /**
+   * Deletes activity types that are not needed by T&W.
```
T&W?

reneolivo replied:
Tasks and Workflows. That's the name of the extension. Should I change it to something else?

mickadoo replied:
Ah, I'd never heard of T&W, apparently it's a thing and I'm just stuck in the past. Never mind my comment

mickadoo commented on uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php:
```diff
@@ -756,15 +794,66 @@ public function upgrade_1033() {
+      'CiviPledge',
+    ]);
+
+    $this->_uninstallActivitiesByLabel([
```
`Label` is usually editable, whereas `name` is supposed to be more like an identifier, can we change this to use names instead of labels?

mickadoo commented on uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php:
```diff
@@ -826,25 +915,59 @@ private function _installActivityTypes($component, array $types) {
+   */
+  private function _uninstallAllComponentActivities(array $componentNames) {
+    $result = civicrm_api3('OptionValue', 'get', [
+      'sequential' => 1,
```
you don't need sequential here

mickadoo commented on uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php:
```diff
@@ -826,25 +915,59 @@ private function _installActivityTypes($component, array $types) {
+    $result = civicrm_api3('OptionValue', 'get', [
+      'sequential' => 1,
+      'option_group_id' => 'activity_type',
+      'component_id' => [ 'IN' => $componentNames ],
```
extra space inside the array brackets

mickadoo commented:
@reneolivo are we sure that these activity types aren't hard-coded in core?

reneolivo commented:
> @reneolivo are we sure that these activity types aren't hard-coded in core?

Good question @mickadoo here's my investigation:

* The activity types for the components are only used inside their respective extensions. These are not relevant for T&W or CiviHR.
* The activities that are being removed by name are part of one of the extensions mentioned above, it's just that they don't have a component_id property. For example, `Downloaded Invoice` and `Emailed Invoice` are part of the CiviContribute extension ([see an example usage here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Form/Task/Invoice.php#L567-L571)).
* None of the `CiviCase` activity types being removed are being used by core.
* The `CiviDocument` activity types being removed are garbage introduced by the sample data and which may be present in current installations.

mickadoo commented on uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php:
```diff
@@ -832,6 +832,7 @@ public function upgrade_1035() {
 
     $this->_uninstallActivityTypes('CiviCase', [
       'Background Check',
+      'Background_Check',
```
Are `Background_Check` and `Background Check` separate activity types?

reneolivo replied:
yes they are. 
![upload](https://user-images.githubusercontent.com/1642119/39241481-0d2f0082-4855-11e8-8407-c76e6ac79412.png)

It seems it got repeated on test sites, not sure why it's like that.

mickadoo replied:
Yeah, I can see on staging too. There's also an activity type of the same name for CiviTask. Anyway, this is fine - it makes sense now

reneolivo commented:
Added the following changes:

* Fixed a missing `return TRUE` statement for one of the upgraders. Without this, the upgraders would not fully run when using the interface.
* There were two activity types with the name `Background Check` and `Background_check`, the last one was missing.
